### PR TITLE
Create a Type Ahead Feature in JavaScript using Fetch and Regex

### LINKED ISSUE #5869: [Languages] Create a Type Ahead Feature in JavaScript  using Fetch and  Regex
#### Proposed title of the article
[Languages] Create a Type Ahead Feature in JavaScript using Fetch and Regex

#### Proposed article introduction
Designing a website with an intuitive user interface is one of the goals/challenges of a web developer.  This tutorial shows how to create a search suggestion for a given search paradigm. What this tutorial does is search a list of States in America from the  [Cites.Json file](https://gist.githubusercontent.com/Miserlou/c5cd8364bf9b2420bb29/raw/2bf258763cdddd704f8ffd3ea9a3e81d25e2c6f6/cities.json), the highlight the typed text when displaying the search result to the user.

To get the Data (i.e. the American state information)  from the JSON file, the fetch() method was used.  The fetch() method provides an easy way to get data to the page asynchronously. Just like XMLhttprequest, Fetch allow the page to load data from an external source without having to reload the page. But the main difference is that Fetch() uses Promises, which enables a simpler and cleaner API, avoiding callback hell.

#### Key takeaways 
The most important things the reader should understand in this tutorial are:
- Highlighting Part of a list element
- More insight on the keyup eventlistener in JS
- A better understanding of regex
- Using Array filter in JS

### PR DESCRIPTION
**Pull request by gmcodes20**
#### Formatting and structure checks
- [X] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [X] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [X] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [X] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [X] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [X] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [X] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [X] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [X] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [X] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [X] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [X] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [X] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [X] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [X] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [X] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

#### Topic suggestion this closes

This closes #5869

### CODE REVIEW

**Comment by srishilesh** [comment]
Hey @gmcodes20 ,

Thanks for submitting this PR.

Your PR has entered our reviewing queue. Please wait until a peer-reviewer pings you with the changes to be made.

Meanwhile, we suggest you to go through other approved PRs/published articles to see if you have missed out anything, which might avoid a potential PR closure.

Thanks.

**Comment by LinusMuema** [comment]
plagiarism check done ✅

File: content/articles/create-an-ajax-type-ahead-feature-for-a-website/index.md [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+topics: [Languages]
+excerpt_separator: <!--more-->
+images:
+
+  - url: /engineering-education/create-an-ajax-type-ahead-feature-for-a-website/hero.jpg
+    alt: Create a Type Ahead Feature in JavaScript using Fetch and Regex Hero Image
+---
+Part of the job of a front-end developer is to create intuitive user interfaces that deliver the best user experience to website users. Part of getting this job done is by creating search suggestions or what is known as the type-ahead feature for websites/webpages.
+<!--more-->
+This tutorial will explain how to create an **ajax type-ahead feature** for your website using the [cities.json file](). This file contains information about states and cities in America. What this feature does is, if a user types `name`, every state and city that matches the word name pops up.
```
Comment (Dawe-7) [inline content/articles/create-an-ajax-type-ahead-feature-for-a-website/index.md:19]: Add the missing link here.

File: content/articles/create-an-ajax-type-ahead-feature-for-a-website/index.md [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+author: godwin-martins
+date: 2022-02-09T00:00:00-10:50
+topics: [Languages]
+excerpt_separator: <!--more-->
+images:
+
+  - url: /engineering-education/create-an-ajax-type-ahead-feature-for-a-website/hero.jpg
+    alt: Create a Type Ahead Feature in JavaScript using Fetch and Regex Hero Image
+---
+Part of the job of a front-end developer is to create intuitive user interfaces that deliver the best user experience to website users. Part of getting this job done is by creating search suggestions or what is known as the type-ahead feature for websites/webpages.
```
Comment (Dawe-7) [inline content/articles/create-an-ajax-type-ahead-feature-for-a-website/index.md:17]: I feel the start of your article is not very intuitive, kindly rephrase it. You could start by saying this 
` Creating intuitive user interfaces that deliver the best user experience to website users is part of the job of a front-end developer`.   

Kindly avoid being repetitive - the word `part`.

File: content/articles/create-an-ajax-type-ahead-feature-for-a-website/index.md [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+<!--more-->
+This tutorial will explain how to create an **ajax type-ahead feature** for your website using the [cities.json file](). This file contains information about states and cities in America. What this feature does is, if a user types `name`, every state and city that matches the word name pops up.
+
+### Prerequisites
+To follow along with this tutorial, the reader should have:
+- Basic HTML and CSS knowledge.
+- ES6 JavaScript/Asynchronous JavaScript.
+- A Code editor.
```
Comment (Dawe-7) [inline content/articles/create-an-ajax-type-ahead-feature-for-a-website/index.md:25]:
```suggestion
- A code editor.
```

File: content/articles/create-an-ajax-type-ahead-feature-for-a-website/index.md [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+
+- Now, we need to fetch our data from the endpoint and assign the returned value to the empty array we created.
+
+```js
+fetch(dataEndpoint)
+  .then((blobdata) => blobdata.json())
+  .then((data) => cities.push(...data));
+```
+
+> The blobdata object, in turn, does not contain the actual JSON response body but is instead a representation of the entire HTTP response. To extract the JSON body content from the Response object, we use the `JSON()` method, which returns a second promise that resolves with the result of parsing the response body text as JSON. -[MDN](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch)
```
Comment (Dawe-7) [inline content/articles/create-an-ajax-type-ahead-feature-for-a-website/index.md:247]: Is the object supposed to be `blobdata` or `BLOB data`. Could you break down the second sentence here as it is too long. Also, add the MDN link properly.

File: content/articles/create-an-ajax-type-ahead-feature-for-a-website/index.md [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+
+![Blobdata output](./returned-blobdata.png)
```
Comment (Dawe-7) [inline content/articles/create-an-ajax-type-ahead-feature-for-a-website/index.md:249]:
```suggestion
![Blobdata output](/engineering-education/create-an-ajax-type-ahead-feature-for-a-website/returned-blobdata.png)
```
Use the same format for your other images as well. Also, if your images are from third party sources/ are not screenshots, add the image source links.

File: content/articles/create-an-ajax-type-ahead-feature-for-a-website/index.md [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+_blobdata returns entire HTTP response_
+
+Here are a few things to note:
+- We defined the empty cities array using a `const` declaration, so we cannot reassign data into it. We can only declare the cities array with a `let` keyword.
+- Alternatively, you can push and [spread the data into the cities array](https://www.samanthaming.com/tidbits/92-6-use-cases-of-spread-with-array/).
+- Create a function called `getMatches` that takes two arguments, the `wordToMatch`and the `cities` array.
+- In the function, we return a filtered subset of the `cities` array and check to see if what the user typed is returned. To check if what the user typed is inside the array, we will [use Regex](https://www.w3schools.com/js/js_regexp.asp) to check if the searched word match any state or city in the `cities` array.
```
Comment (Dawe-7) [inline content/articles/create-an-ajax-type-ahead-feature-for-a-website/index.md:256]:
```suggestion
 In the function above, we return a filtered subset of the `cities` array and check if what the user typed is returned. To check if what the user typed is inside the array, we will use [Regex](https://www.w3schools.com/js/js_regexp.asp) to check if the searched word matches any state or city in the `cities` array.
```

File: content/articles/create-an-ajax-type-ahead-feature-for-a-website/index.md [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+- [The complete JavaScript code](#the-complete-javascript-code)
+- [Conclusion](#conclusion)
+
+### How `fetch` works
+`Fetch` like `XMLHttpRequest` provides a method to get resources asynchronously across networks. However, the `Fetch` method gives an easy, cleaner, and more efficient way to handle resources across networks.
+
+This method uses `Promise` to deliver more flexible features to make requests to servers from the web browsers.
+
+_Syntax_
+`Fetch(URL, options)`
```
Comment (Dawe-7) [inline content/articles/create-an-ajax-type-ahead-feature-for-a-website/index.md:44]: Place this method in backticks and identify the language.

**Review by Dawe-7 — CHANGES_REQUESTED** [review]
Nice article @gmcodes20, I've left you a few suggestions and comments. Kindly effect the changes so that we can move on to tech review. Your hero image is too large, make sure it is of the size 600x400 pixels. Resize your other images as well if they are too large.

Also, make sure to cross check your article for readability because there were some readability issues in review.

**Review by Dawe-7 — APPROVED** [review]
Awesome Job on the article @gmcodes20 . Kindly go through the changes made throughout the review process so as to improve on the flow and format of your future articles. Thank you.

**Comment by gmcodes20** [comment]
> Awesome Job on the article @gmcodes20 . Kindly go through the changes made throughout the review process so as to improve on the flow and format of your future articles. Thank you.

Thank you @Dawe-7 , I'll keep that in mind.

**Comment by Kayere** [comment]
Great work gents 🚀.

**Comment by gmcodes20** [comment]
> Great work gents 🚀.

Thank you 😊

**Pull request merged**